### PR TITLE
[mungegithub] Add a new munger that forces people to file or reference flaky tests

### PR DESCRIPTION
**Pull request by brendandburns**
@ixdy @spxtr @ihmccreery 

Ref: https://github.com/kubernetes/kubernetes/issues/20097

### CODE REVIEW

File: mungegithub/mungers/rebuild.go [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+			if obj.HasLabel("lgtm") {
+				if err := obj.RemoveLabel("lgtm"); err != nil {
+					glog.Errorf("unexpected error removing lgtm label: %v", err)
+				}
+			}
+		}
+	}
+}
+
+func isInvalidRebuildComment(comment *githubapi.IssueComment) bool {
```
Comment (lavalamp) [inline mungegithub/mungers/rebuild.go]: Nit: stuff is so much easier to read when named in the positive :)

File: mungegithub/mungers/rebuild.go [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+	"github.com/spf13/cobra"
+)
+
+// RebuildMunger looks for situations where a someone has asked for an e2e rebuild, but hasn't provided
+// an issue
+type RebuildMunger struct{}
+
+var (
+	buildMatcher = regexp.MustCompile("@k8s-bot\\s+test\\s+this.*")
+	issueMatcher = regexp.MustCompile("\\s+(?:github\\s+)?issue\\:?\\s+#(?:IGNORE|[0-9]+)")
```
Comment (lavalamp) [inline mungegithub/mungers/rebuild.go]: Can `flake: #12345` be allowed? "github issue" and "issue" seem redundant.

**Comment by lavalamp** [comment]
Minor suggestions.

File: mungegithub/mungers/rebuild.go [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+	githubapi "github.com/google/go-github/github"
+	"github.com/spf13/cobra"
+)
+
+// RebuildMunger looks for situations where a someone has asked for an e2e rebuild, but hasn't provided
+// an issue
+type RebuildMunger struct{}
+
+var (
+	buildMatcher = regexp.MustCompile("@k8s-bot\\s+test\\s+this.*")
```
Comment (spxtr) [inline mungegithub/mungers/rebuild.go]: `@k8s-bot e2e test this` and `@k8s-bot unit test this` also trigger the respective PR Jenkins jobs.

File: mungegithub/mungers/rebuild.go (hunk 1 above)
Comment (brendandburns) [inline mungegithub/mungers/rebuild.go]: done.

File: mungegithub/mungers/rebuild.go (hunk 2 above)
Comment (brendandburns) [inline mungegithub/mungers/rebuild.go]: done.

File: mungegithub/mungers/rebuild.go (hunk 3 above)
Comment (brendandburns) [inline mungegithub/mungers/rebuild.go]: done.

**Comment by brendandburns** [comment]
comments addressed, ptal.

thanks
--brendan

File: mungegithub/mungers/rebuild_test.go [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+		{
+			value:         "@k8s-bot test this please github issue #",
+			expectMissing: true,
+		},
+		{
+			value:         "@k8s-bot test this please github issue IGNORE",
+			expectMissing: true,
+		},
+		{
+			value:         "@k8s-bot test this please issue: #IGNORE",
```
Comment (lavalamp) [inline mungegithub/mungers/rebuild_test.go]: Nit: flake test case?

**Comment by lavalamp** [comment]
LGTM but please add one more test case ;)

Merge yourself when you're ready to push.

**Comment by eparis** [comment]
removed LGTM while I check out a theory...

**Comment by eparis** [comment]
So every 10 minutes the bot is going to spam a new comment complaining about the old comment and the only way to get out of the situation will be for something to delete the old comment....

**Comment by lavalamp** [comment]
Heh, I guess that's what I get for reviewing before finishing the coffee. I guess I was assuming it was going to delete the offending comment...

**Comment by eparis** [comment]
Can we get @ixdy to put the same requirements on the actual tests? So this is enforced?

**Comment by brendandburns** [comment]
ah, @eparis good catch, I will look for the added comment, and only add it if it's not there...

**Comment by eparis** [comment]
deleting the offending comment is actually really easy...  If you'd rather do that @brendandburns 

I can write the tee tiny github helper so you just have to call comment.Delete()....

**Comment by eparis** [comment]
I fact https://github.com/kubernetes/contrib/pull/405 fixes your ListComments helper and write a DeleteComment helper....

**Comment by eparis** [comment]
I'll also say that I'm worried about the number of API tokens we are going to use. We get 5000/hr and usually have like 2500 to spare. If every 10 minutes 100 PRs get comments this will only eat 600 tokens/hr, which should be fine. But I have no idea how many PRs get comments every 10 minutes.

We might need to slow the bot down to every 15 minutes or something if we are chewing through too many tokens.

**Comment by brendandburns** [comment]
@eparis good suggestion wrt to delete.  I went ahead and added that in.  ptal.

thanks!
--brendan

**Comment by brendandburns** [comment]
(and I will rebase on #405 once that's in)

File: mungegithub/mungers/rebuild.go [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+
+	comments, err := obj.GetComments(*obj.Issue.Number)
+	if err != nil {
+		glog.Errorf("unexpected error getting comments: %v", err)
+	}
+
+	for ix := range comments {
+		comment := &comments[ix]
+		// Skip all robot comments
+		for _, robot := range []string{"k8s-bot", "googlebot"} {
```
Comment (eparis) [inline mungegithub/mungers/rebuild.go]: k8s-merge-robot?

**Comment by eparis** [comment]
I say update the one more robot and merge it. I'll rebase #405

File: mungegithub/mungers/rebuild.go [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+				glog.V(4).Infof("Skipping comment by robot %s: %s", robot, *comment.Body)
+				continue
+			}
+		}
+		if isRebuildComment(comment) && rebuildCommentMissingIssueNumber(comment) {
+			if err := obj.DeleteComment(*comment.ID); err != nil {
+				glog.Errorf("Error deleting comment: %v", err)
+				continue
+			}
+			err := obj.WriteComment(fmt.Sprintf("@%s an issue is required for any manual rebuild.  Expecting comment of the form 'github issue: #<number>'", *comment.User.Login))
```
Comment (ixdy) [inline mungegithub/mungers/rebuild.go]: should we also link to the list of known flakes? i.e. https://github.com/kubernetes/kubernetes/issues?q=is%3Aissue%20label%3Akind%2Fflake ?

**Comment by eparis** [comment]
going to merge and will cleanup in additional PRs

**Pull request merged**